contracts/vanchor/src/state.rs

```diff
@@ -0,0 +1,313 @@
+
+// Struct to save the use of "nullifiers"
+pub const NULLIFIERS: Map<Vec<u8>, bool> = Map::new("used_nullifers");
+
+// "Poseidon hasher"
+pub const POSEIDON: Item<Poseidon> = Item::new("poseidon");
+
+// "VAnchorVerifier"
+pub const VANCHORVERIFIER: Item<VAnchorVerifier> = Item::new("vanchor_verifier");
+
+// MerkleTree
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct MerkleTree {
+    pub levels: u32,
+    pub current_root_index: u32,
+    pub next_index: u32,
+}
+
+// MerkleTree "filled_subtrees"
+pub const FILLED_SUBTREES: Map<String, [u8; 32]> = Map::new("filled_subtrees");
+
+pub fn save_subtree(store: &mut dyn Storage, k: u32, data: &[u8; 32]) -> StdResult<()> {
+    FILLED_SUBTREES.save(store, k.to_string(), data)
+}
+
+pub fn read_subtree(store: &dyn Storage, k: u32) -> StdResult<[u8; 32]> {
+    FILLED_SUBTREES.load(store, k.to_string())
+}
+
+// MerkleTree Roots
+pub const MERKLEROOTS: Map<String, [u8; 32]> = Map::new("merkle_roots");
+
+pub fn save_root(store: &mut dyn Storage, k: u32, data: &[u8; 32]) -> StdResult<()> {
+    MERKLEROOTS.save(store, k.to_string(), data)
+}
+
+pub fn read_root(store: &dyn Storage, k: u32) -> StdResult<[u8; 32]> {
+    MERKLEROOTS.load(store, k.to_string())
+}
+
+impl MerkleTree {
+    fn hash_left_right(
+        &self,
+        hasher: Poseidon,
+        left: [u8; 32],
+        right: [u8; 32],
+    ) -> Result<[u8; 32], ContractError> {
+        let inputs = vec![left, right];
+        hasher.hash(inputs).map_err(|_e| ContractError::HashError)
+    }
+
+    pub fn insert(
+        &mut self,
+        hasher: Poseidon,
+        leaf: [u8; 32],
+        store: &mut dyn Storage,
+    ) -> Result<u32, ContractError> {
+        let next_index = self.next_index;
+        assert!(
+            next_index != 2u32.pow(self.levels as u32),
+            "Merkle tree is full"
+        );
+
+        let mut current_index = next_index;
+        let mut current_level_hash = leaf;
+        let mut left: [u8; 32];
+        let mut right: [u8; 32];
+
+        for i in 0..self.levels {
+            if current_index % 2 == 0 {
+                left = current_level_hash;
+                right = zeroes::zeroes(i);
+                save_subtree(store, i, &current_level_hash)?;
+            } else {
+                left = read_subtree(store, i).map_err(|_| ContractError::HashError)?;
+                right = current_level_hash;
+            }
+
+            current_level_hash = self.hash_left_right(hasher.clone(), left, right)?;
+            current_index /= 2;
+        }
+
+        let new_root_index = (self.current_root_index + 1) % ROOT_HISTORY_SIZE;
+        self.current_root_index = new_root_index;
+        save_root(store, new_root_index, &current_level_hash)?;
+        self.next_index = next_index + 1;
+        Ok(next_index)
+    }
+
+    pub fn is_known_root(&self, root: [u8; 32], store: &dyn Storage) -> bool {
+        // if root == [0u8; 32] {
```

Comment:
Can we uncomment this?

Reply:
The reason I commented this line is that the very first transact(deposit) is going with zero root.